devito/types/basic.py

```diff
@@ -652,6 +640,8 @@ def __new__(cls, *args, **kwargs):
 
         # Store new instance in symbol cache
         Cached.__init__(newobj, newcls)
+        # Cache with indices as well to avoid recreating the base function
+        Cached.__init__(newobj, (newcls, *indices), alias=newcls)
```

Comment:
calling `__init__` twice is really weird... you also end up changing state (`_cache_key_hash` in this case)!

see my other comment below

Reply:
Well you "change" the state by replacing it with itself so it's not actually changing but I get your point.